content/blog/2018/02/2018-02-22-cheetah.adoc

```diff
@@ -0,0 +1,134 @@
+** Another common pattern is a "summary" object containing data from many branches (logs, results, or statistics). Often this is visible because you'll be adding to it often via an add/append or `Map.put()` operations.
+** Large arrays of data or Maps of configuration information are another common example of this situation.
+* No, if your Pipelines spend almost all their time waiting for a few shell/batch steps to finish.  This ISN'T a magic "go fast" button for everything!
+* No, if Pipelines are writing massive amounts of data to logs (logging is unchanged).
+* No, if you are not using Pipelines, or your system is loaded down by other factors.
+* No, if you don't enable higher-performance modes for pipelines.
+
+== Other Goodies
+
+* Users can now set a job property so that individual Pipelines fail cleanly rather than resuming upon restarting the master. 
+
+* We've reduced classloading and reflection quite significantly, which improves scaling and reduces CPU use:
+
+image:/images/post-images/2018-02-22-cheetah/classloading.png[role="center"]
+
+* Script Security has gotten optimizations to reduce the performance overhead of Sandbox mode and elliminate lock contention so Pipeline multithreads better. 
+
+* Pipeline Step data uses up less space on disk (regardless of the durability setting) - this should be 30% smaller.  Assume it's a few MB per 1000 steps - but for every build after the change. 
+
+* Even in the low-performance/high-durability modes, some redundant writes have been removed, which decreases the number of writes by 10-20%.
+
+== How Did You Do It?
+
+That's probably material for another blog post or https://www.cloudbees.com/jenkinsworld[Jenkins World talk].
+
+The short answer is: first we built a tool to simulate a full production environment and provide detailed metrics collection at scale.  Then we profiled Jenkins to identify bottlenecks and attacked them.  Rinse and repeat.
+
+== What Next?
+
+The next big change, which I'm calling Cheetah Part 2 is to address Pipeline's logging. For every Step run, Pipeline writes one or more small log files. These log files are then copied into the build log content, but are retained to make it possible to easily fetch logs for each step. 
+
+This copying process means every log line is written twice, greatly reducing perforance, and writing to many small files is orders of magnitude slower than appending to one big log file.
+
+We're going to remove this duplication and data fragmentation and use a more efficient mechanism to find per-step logs. This should further improve the ability to run Pipelines on NFS mounts and hard-drive-backed storage, and should significantly improve performance at scale.
+
+Besides this, there's a variety of different tactical improvements to improve scaling behavior and reduce resource needs.
```

Comment:
I suggest closing out by referencing your previous blog post once again and highlighting one of the things highlighted in that post: there is no performance silver bullet, so first measure and then start to tune

Reply:
I've already called out out where this change and the performance-optimized mode does-and-doesn't help for users.   Adding this would undersell the new value being offered here -- I've done the work to measure-and-tune with a degree of rigor that can't be matched unless others use Hydra.  The point of this is to *reduce* the burden on the user, not increase it.